### PR TITLE
type/compatibility: check `KEY` option for generated column.

### PR DESCRIPTION
**Pull request by pingyu**
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->
check `KEY` option for generated column in `create table`

TiDB:
```sql
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NULL, PRIMARY KEY (id));
ERROR 1068 (42000): Multiple primary key defined
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NOT NULL, PRIMARY KEY (id));
ERROR 1068 (42000): Multiple primary key defined
```

MySQL:
```sql
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NULL, PRIMARY KEY (id));
ERROR 3106 (HY000): 'Defining a virtual generated column as primary key' is not supported for generated columns.
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NOT NULL, PRIMARY KEY (id));
ERROR 3106 (HY000): 'Defining a virtual generated column as primary key' is not supported for generated columns.
mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 5.7.25-0ubuntu0.18.04.2 |
+-------------------------+
1 row in set (0.00 sec)

mysql> 
```

related issue: #9464 

### What is changed and how it works?
add checking of `KEY` option on generated column for `CREATE TABLE`

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->
 - Manual Test
```sql
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NULL, PRIMARY KEY (id));
ERROR 3106 (HY000): 'Defining a virtual generated column as primary key' is not supported for generated columns.
mysql> CREATE TABLE t1 (id INT NOT NULL,
    ->   c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NOT NULL, PRIMARY KEY (id));
ERROR 3106 (HY000): 'Defining a virtual generated column as primary key' is not supported for generated columns.
mysql> select TiDB_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| TiDB_version()                                                                                                                                                                                                                                                                                                                             |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-144-g560e8cfe5-dirty
Git Commit Hash: 560e8cfe57d102d52918fd609e4e865ebdfb7159
Git Branch: master
UTC Build Time: 2019-03-03 05:20:02
GoVersion: go version go1.11.2 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
```

 - Unit test

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -417,13 +422,24 @@ func isTableAliasDuplicate(node ast.ResultSetNode, tableAliases map[string]inter
 	return nil
 }
 
-func isPrimary(ops []*ast.ColumnOption) int {
+func checkColumnOptions(ops []*ast.ColumnOption) (int, error) {
+	isPrimary := 0
+	isGenerated := 0
+
```
Comment (imtbkcat) [inline planner/core/preprocess.go]: `isPrimary, isGenerated := 0, 0` is better

Comment (pingyu) [reply planner/core/preprocess.go]: @imtbkcat PTAL. Thanks.

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -331,7 +331,12 @@ func (p *preprocessor) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
 			p.err = errors.Trace(err)
 			return
 		}
-		countPrimaryKey += isPrimary(colDef.Options)
+		isPrimary, err := checkColumnOptions(colDef.Options)
+		if err != nil {
+			p.err = errors.Trace(err)
```
Comment (lamxTyler) [inline planner/core/preprocess.go]: No need to trace the error anymore.

Comment (pingyu) [reply planner/core/preprocess.go]: @lamxTyler PTAL. Thanks.

**Comment by lamxTyler** [comment]
The test failed because the sql `create table t (a DOUBLE NULL, b_sto DOUBLE GENERATED ALWAYS AS (a + 2) STORED UNIQUE KEY NOT NULL PRIMARY KEY);` could be successfully executed in MySQL, while it raises an error in this PR.

**Comment by pingyu** [comment]
Let me see see.

**Comment by pingyu** [comment]
Bug fixed. Primary key for "stored generated column" is permitted.
@lamxTyler PTAL. Thanks.

**Comment by pingyu** [comment]
```sql
mysql> create table t (a DOUBLE NULL, b_sto DOUBLE GENERATED ALWAYS AS (a + 2) STORED UNIQUE KEY NOT NULL PRIMARY KEY);
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                      |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` double DEFAULT NULL,
  `b_sto` double GENERATED ALWAYS AS (`a` + 2) STORED NOT NULL,
  UNIQUE KEY `b_sto` (`b_sto`),
  PRIMARY KEY (`b_sto`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_version();
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                                                    |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta.1-48-g0561316ac-dirty
Git Commit Hash: 0561316ac18ab9a96b036f8f3de2daac86edeace
Git Branch: check_key_for_virtual_column
UTC Build Time: 2019-04-05 07:34:29
GoVersion: go version go1.11.2 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 

```

**Comment by lamxTyler** [comment]
@pingyu Please resolve the conflicts.

**Comment by pingyu** [comment]
@lamxTyler Conflicts resolved. PTAL. Thanks.

**Pull request merged**